component/topsql/subscriber/scraper.go

```diff
@@ -255,48 +268,47 @@ func (bo *backoffScrape) reconnect() {
 		case "tidb":
 			client := tipb.NewTopSQLPubSubClient(conn)
 			bo.client = client
-			if bo.stream, err = client.Subscribe(bo.ctx, &tipb.TopSQLSubRequest{}); err != nil {
+			stream, err := client.Subscribe(bo.ctx, &tipb.TopSQLSubRequest{})
+			if err != nil {
+				log.Warn("failed to call Subscribe", zap.Any("component", bo.component), zap.Error(err))
+				bo.retryTimes += 1
```

Comment:
Just a suggestion. The current retry logic and business logic are coupled together, and each retry logic must be implemented by itself. Maybe we can implement a set of general tools in the `util` package, like:
```go
func WithRetry(times int, duration time.Duration, f func() error) error {
	var err error
	for retry := 0; retry <= times; retry++ {
		err = f()
		if err == nil {
			break
		}
		if retry < times {
			time.Sleep(duration)
		}
	}
	return err
}

func WithRetryBackoff(times int, f func() error) error {
	duration := time.Second
	var err error
	for retry := 0; retry <= times; retry++ {
		err = f()
		if err == nil {
			break
		}
		if retry < times {
			time.Sleep(duration)
			duration *= 2
		}
	}
	return err
}
```

Then we can simply do:
```go
err := WithRetry(3, time.Second, func() error {
    /* ... */
})
if err != nil {
    /* ... */
}
```

Reply:
I will have a try.